src/eval.rs

```diff
@@ -1,26 +1,998 @@
+            } else if v < lower {
+                Ok(Value::Num(lower, range))
+            } else {
+                Ok(Value::Num(v, range))
+            }
+        }
+        Overflow::Wraparound => Ok(Value::Num(
+            ((v - lower).abs() % (upper - lower) + lower),
+            Range {
+                start: lower,
+                end: upper,
+            },
+        )),
+        _ => Err(RuntimeError::InvalidExpression),
+    };
+    return val;
+}
 
 // eval. expr.
+fn eval_expr(expr: &Expr, env: &Env, program: &Program) -> Result<Value, RuntimeError> {
+    match expr {
+        Expr::Num(n, Type::NumT(range)) => cast(*n, range.clone(), Overflow::Wraparound),
+        Expr::Bool(b) => Ok(Value::Bool(*b)),
+        Expr::Sym(symbol) => Ok(Value::Sym(Symbol(*symbol))),
+
+        Expr::Var(id) => Ok(env.get(id).unwrap().clone()),
+
+        Expr::BinOp { lhs, op, rhs } => {
+            let left = eval_expr(lhs, env, &program)?;
+            let right = eval_expr(rhs, env, &program)?;
+
+            match (left, right) {
+                (Value::Num(l, l_range), Value::Num(r, range_ignore)) => match op {
+                    // numerical return
+                    BOp::Add => cast(l + r, l_range, Overflow::Wraparound),
+                    BOp::Sub => cast(l - r, l_range, Overflow::Wraparound),
+                    BOp::Mul => cast(l * r, l_range, Overflow::Wraparound),
+                    BOp::Div => {
+                        if r == 0 {
+                            Err(RuntimeError::DivisionbyZero)
+                        } else {
+                            cast(l / r, l_range, Overflow::Wraparound)
+                        }
+                    }
+                    BOp::Rem => {
+                        if r == 0 {
+                            Err(RuntimeError::ModulusByZero)
+                        } else {
+                            cast(l % r, l_range, Overflow::Wraparound)
+                        }
+                    }
+                    BOp::Shl => cast(l << r, l_range, Overflow::Wraparound),
+                    BOp::Shr => cast(l >> r, l_range, Overflow::Wraparound),
+
+                    // boolean return
+                    BOp::Lt => Ok(Value::Bool(l < r)),
+                    BOp::Lte => Ok(Value::Bool(l <= r)),
+                    BOp::Eq => Ok(Value::Bool(l == r)),
+                    BOp::Ne => Ok(Value::Bool(l != r)),
+
+                    _ => Err(RuntimeError::InvalidOperand),
+                },
+
+                (Value::Bool(l), Value::Bool(r)) => match op {
+                    BOp::And => Ok(Value::Bool(l && r)),
+                    BOp::Or => Ok(Value::Bool(l || r)),
+
+                    _ => Err(RuntimeError::InvalidOperand),
+                },
+                _ => Err(RuntimeError::TypeError),
+            }
+        }
+
+        Expr::UOp { op, inner } => {
+            let left = eval_expr(inner, env, &program)?;
+
+            match left {
```

Comment:
you can combine this match with inner matches. For example:

```rust
match (left, op) {
  (Value::Num(l, range), UOp::Negate) => ...,
  ....
}
```

Also, `InvalidOperand` is just `TypeError` in this case, no need to have separate types for them.